### PR TITLE
Cartesian control

### PR DESCRIPTION
**Pull request by fmauch**
This PR adds Cartesian streaming control and controllers to the driver.

In analogy to the JointPosition and JointVelocity interface, a Pose and Twist interface is added. We have three controllers added as  examples:
 - `twist_controller`: publish a cartesian twist to the robot's tcp
 - `pose_based_cartesian_traj_controller`: Cartesian trajectory execution using a pose interface
 - `joint_based_cartesian_traj_controller`: Cartesian trajectory execution using an IK and a joint position interface

### CODE REVIEW

File: ur_robot_driver/package.xml [hunk 1]
```diff
@@ -41,10 +41,12 @@
   <depend>tf2_msgs</depend>
   <depend>tf</depend>
   <depend>trajectory_msgs</depend>
+  <depend>twist_controller</depend>
```
Comment (stefanscherzinger) [inline ur_robot_driver/package.xml:44]: Would `exec_depend` be sufficient here?

File: ur_robot_driver/src/hardware_interface.cpp [hunk 2]
```diff
@@ -746,6 +793,15 @@ void HardwareInterface::doSwitch(const std::list<hardware_interface::ControllerI
         {
           cartesian_forward_controller_running_ = false;
         }
+        if (resource_it.hardware_interface == "ros_controllers_cartesian::TwistCommandInterface")
+        {
+          twist_controller_running_ = false;
+        }
+        if (resource_it.hardware_interface == "ros_controllers_cartesian::PoseCommandInterface")
+        {
+          ROS_INFO_STREAM("Stopping pose controller");
```
Comment (stefanscherzinger) [inline ur_robot_driver/src/hardware_interface.cpp:802]: This seems to be the only controller type that gets an info output. Is this intention?

File: ur_robot_driver/src/hardware_interface.cpp [hunk 3]
```diff
@@ -784,6 +840,15 @@ void HardwareInterface::doSwitch(const std::list<hardware_interface::ControllerI
         {
           cartesian_forward_controller_running_ = true;
         }
+        if (resource_it.hardware_interface == "ros_controllers_cartesian::TwistCommandInterface")
+        {
+          twist_controller_running_ = true;
+        }
+        if (resource_it.hardware_interface == "ros_controllers_cartesian::PoseCommandInterface")
+        {
+          ROS_INFO_STREAM("Starting pose controller");
```
Comment (stefanscherzinger) [inline ur_robot_driver/src/hardware_interface.cpp:849]: Same here... Output wanted?

File: ur_robot_driver/test/integration_test.py [hunk 4]
```diff
@@ -87,6 +91,15 @@ def init_robot(self):
                 "Could not reach joint passthrough controller action. Make sure that the driver is actually running."
                 " Msg: {}".format(err))
 
+        self.cartesian_trajectory_client = actionlib.SimpleActionClient(
+            'follow_cartesian_trajectory', FollowCartesianTrajectoryAction)
+        try:
```
Comment (stefanscherzinger) [inline ur_robot_driver/test/integration_test.py:96]: I guess `wait_for_server` does not throw but [returns false on timeout](https://docs.ros.org/en/noetic/api/actionlib/html/simple__action__client_8py_source.html#l00067).
Maybe simply use:

```python
        if not self.cartesian_trajectory_client.wait_for_server(timeout):
            self.fail("Could not reach cartesian controller action. Make sure that the driver is actually running.")

```

**Review by stefanscherzinger — APPROVED** [review]
Looks correct to me. Left some minor suggestions.

File: ur_robot_driver/package.xml (hunk 1 above)
Comment (fmauch) [reply ur_robot_driver/package.xml:44]: yes, that should be sufficient

File: ur_robot_driver/src/hardware_interface.cpp (hunk 2 above)
Comment (fmauch) [reply ur_robot_driver/src/hardware_interface.cpp:802]: That seems to be a leftover from development. I will remove that.

File: ur_robot_driver/src/hardware_interface.cpp (hunk 3 above)
Comment (fmauch) [reply ur_robot_driver/src/hardware_interface.cpp:849]: Same as above

File: ur_robot_driver/test/integration_test.py (hunk 4 above)
Comment (fmauch) [reply ur_robot_driver/test/integration_test.py]: Thanks for spotting this. `ServiceProxy.wait_for_service` [throws an exception](https://github.com/ros/ros_comm/blob/noetic-devel/clients/rospy/src/rospy/impl/tcpros_service.py#L93), while  [`ActionClient.wait_for_server`](https://docs.ros.org/en/noetic/api/actionlib/html/action__client_8py_source.html#l00590) does not.

Comment (fmauch) [reply ur_robot_driver/test/integration_test.py]: Resolved in 65e55ac

**Pull request merged**